### PR TITLE
Fix: replace incorrect type predicate with common Geometry Node interface

### PR DESCRIPTION
**Pull request by JordanLane-Veerum**
Hey guys, great work on this repo. I've enjoyed playing around with the example and the codebase. 

Recently I encountered an incorrect type-predicate in the `isGeometryNode` function. When running the priorityQueue loop, an OctreeGeometryNode would be incorrectly asserted as a PointCloudOctreeGeometryNode which could cause subtle typing bugs. 

To fix this I created a common interface IPointCloudGeometryNode that extends the IPointCloudTreeNode interface and adds the geometry specific attributes. 

three-loader-debug

### CODE REVIEW

File: src/loading2/octree-geometry-node.ts [hunk 1]
```diff
@@ -15,8 +15,9 @@ export class OctreeGeometryNode implements IPointCloudTreeNode {
 
 	loaded: boolean = false;
 	loading: boolean = false;
+	failed: boolean = false;
 	parent: OctreeGeometryNode | null = null;
-	geometry: BufferGeometry | null = null;
+	geometry: BufferGeometry | undefined;
```
Comment (pjurczynski) [inline src/loading2/octree-geometry-node.ts:20]: Could you explain a bit why we're changing the type here?

File: src/loading2/octree-geometry-node.ts [hunk 2]
```diff
@@ -67,15 +68,18 @@ export class OctreeGeometryNode implements IPointCloudTreeNode {
...
 
-		if (this.octreeGeometry.loader) {
-			this.octreeGeometry.loader.load(this);
+		if (!this.octreeGeometry.loader) {
+			this.loading = false;
+			this.failed = true;
+			return Promise.reject(`Loader not initialized for ${this.name}`);
 		}
+
+		return this.octreeGeometry.loader.load(this);
```
Comment (pjurczynski) [inline src/loading2/octree-geometry-node.ts:82]: these changes seem irrelevant to the proposed changes in the PR.

**Review by pjurczynski — CHANGES_REQUESTED** [review]
Thanks for taking the time to look into the library and propose an improvement. It's very much appreciated.

I got some minor questions for you regarding the change.

File: src/loading2/octree-geometry-node.ts (hunk 1 above)
Comment (JordanLane-Veerum) [reply src/loading2/octree-geometry-node.ts:20]: The main reason is the `toTreeNode` method of the `PointCloudOctree` class. 

I changed the `toTreeNode` method to use the updated `IPointCloudGeometryNode` interface. Then in that method, when calling the `Points()` constructor we can pass in undefined geometry which allows ThreeJS to still properly initialize that class (which is not possible with null)

File: src/loading2/octree-geometry-node.ts (hunk 2 above)
Comment (JordanLane-Veerum) [reply src/loading2/octree-geometry-node.ts:82]: I implemented these changes as the new IPointCloudGeometryNode interface enforces the `load` function returning a `Promise` type. 

I wanted to make sure that OctreeGeometryNode also properly set the loading and failed attributes properly, however I can move this to separate PR if you would prefer!

File: src/loading2/octree-geometry-node.ts (hunk 1 above)
Comment (pjurczynski) [reply src/loading2/octree-geometry-node.ts:20]: oh, that's nice! Thanks for the explanation.

File: src/loading2/octree-geometry-node.ts (hunk 2 above)
Comment (pjurczynski) [reply src/loading2/octree-geometry-node.ts:82]: It's just fine within the contents of this PR.

**Review by pjurczynski — APPROVED** [review]

**Review by MPanknin — APPROVED** [review]
Looks good to me. Sry for the late response and thanks for the contribution.

**Comment by MPanknin** [comment]
https://www.npmjs.com/package/@pnext/three-loader
